### ISSUE TITLE
Scaling strokes (use Cairo.jl's `stroke_transformed` and `stroke_transformed_preserve`)

### ISSUE THREAD

**Issue by JeffreyPalmer** [JuliaGraphics/Luxor.jl#163]
Hi! Thanks for creating this library - it does a lot of what I am looking for so that was great to find.

I have one question, though. I would like to use this library for my generative artwork, and my normal approach with other libraries is to scale my image coordinate system from 0 to 1 and then use those normalized non-pixel coordinates for everything, relying on all of the primitives to act accordingly when I scale the canvas up for print resolution. However, I noticed that the Cairo.jl library wrapper made the decision to not scale strokes, which is incompatible with my preferred approach. I also noticed that you mentioned this scaling issue directly in your documentation, recommending the creation of a `strokeraw` function that calls the Cairo primitives directly. However, I am a little at a loss in terms of understanding how to do that.

For now, I've been directly overriding both `stroke` and `strokepreserve` to use Cairo.jl's `stroke_transformed` and `stroke_transformed_preserve`, but that seems at odds with your recommendation. 

Since I'm very new to the Julia ecosystem, I'd appreciate a little guidance. Do you have any examples that use the `strokeraw` approach that you mentioned? Am I missing something obvious?

Thanks for your help!

**Comment by cormullion** [comment]
Hi Jeffery! It's been a few years since I wrote that paragraph and I can't currently remember whether I even tested it... 😱

I'll have a look tomorrow morning and see what I can find.

**Comment by JeffreyPalmer** [comment]
Thanks for the explanation. I'll put together a PR with the change that you outlined. This will be a good opportunity to experiment with Julia package development and contribution, as that's not something that I've done yet.

Regarding the coordinate system, this is something that I prefer after working with Solandra (https://github.com/jamesporter/solandra). I also used a similar approach with SVGs, since they can have an internal coordinate system, but I've since stopped using SVGs because they don't scale to the types of images that I create these days.

Regarding terminology, my preference would actually be for unscaled lines to be called that explicitly, as I find the decision made on the Cairo.jl side to override the scaling behavior of a single rendering primitive to be utterly bizarre. So, if the past wasn't an issue it would be something like `:stroke` and `:stroke_unscaled`. Because we don't live in that world, I don't yet have a good idea, but I'll think on it prior to submitting the PR.

As a side note, I have also been implementing some other algorithms that I need, things like Poisson disc sampling and Chaiken corner cutting. I assume there will be more. Are these things that you'd be interested in adding to Luxor?

Thanks again for your help!

**Comment by JeffreyPalmer** [comment]
Okay so I did a little more digging and apparently the convention to not scale strokes comes from Graphics.jl, which defines that behavior. With that in mind it probably make sense to use terminology consistent with the other libraries, so the `:stroke` action should have a `:stroke_transformed` variant, as should the `:stroke_preserve` action.

**Comment by JeffreyPalmer** [comment]
I was thinking about this a little more, and I have an alternative approach I'd like to run by you. Instead of exposing the fact that there are both types of strokes through the addition of several new actions and other functions (e.g., `fillstroke` directly calls `strokepath` and would need a complementary function), what do you think about including a boolean to the drawing constructor that indicates whether strokes should be scaled, and then checking this value within `strokepath` and `stroke_preserve` and calling the appropriate Cairo function? This would add a parameter to the constructor functions and leave the rest of the API basically untouched, which seems a little better than exposing the fact that these variants exist and need to be considered. Thoughts?

**Comment by cormullion** [comment]
Yes, I also thought of that approach. There are lots of places you could have a Boolean switch - ranging from an environment variable to a drawing constructor option. I think the level I fancy most is something like:

```julia
setstrokescale(true)
```

to start scaling strokes, and 

```julia
setstrokescale(false)
```

to stop scaling strokes. With

```julia
setstrokescale()
``` 

returning the current setting. This would probably be initialised to `false` unless otherwise set. (Since there's been no demand for this feature, I suspect it's not going to be used by too many people...)

This would allow you to change the setting whenever you wanted...

As for other additions to Luxor, that would be great! Although I'd point out that  the Julia ecosystem is dominated by frequent package interdependence, perhaps more than with some other languages. (See one of Stefan's "The Unreasonable Effectiveness of Multiple Dispatch" presentations on YouTube...) It's more usual that a new set of functionality is provided to the entire ecosystem as a stand-alone package, rather than built-in to a few more monolithic ones. So increasingly people are using packages from JuliaGeometry and other organizations. So you use Rotations from Rotations.jl, coordinate transformations from CoordinateTransforms.jl, and so on. So this way more people can benefit from the developers' work and they don't have to take all the other stuff on board as well. (That Luxor doesn't do this as much as it should is probably because it predates a fair few...)

It's an interesting question - how much functionality should be added to a package before it gets bloated, or becomes a development burden. I'm not sure of an answer yet. 😃 

>the decision made on the Cairo.jl side to override the scaling behavior of a single rendering primitive to be utterly bizarre.

I think that's overstating things. An equally valid mental model for end users to have is that a 1-point pen will draw 1-point strokes, no matter where on the drawing you're working - real life is often an appropriate metaphor to adopt.

**Comment by JeffreyPalmer** [comment]
Sounds good. Yeah, our mental models are just slightly different and there's nothing wrong with that. :) 

I'll implement this as you outlined and get back to you. Thanks!

**Comment by JeffreyPalmer** [comment]
Also, your comment about encouraging the creation of smaller utility libraries is very interesting. I was planning to create a new library to house these algorithms, but I thought I'd get your thoughts, so the context on the community expectations around the package ecosystem are very helpful. Thanks!

**Comment by JeffreyPalmer** [comment]
First draft PR created here: https://github.com/JuliaGraphics/Luxor.jl/pull/164

Please let me know what I've missed, if you'd like things changed, etc. This is my first Julia PR, so I am sure that I missed/don't understand something.

**Comment by cormullion** [comment]
Closed by #164. Thanks!
